src/librustc_mir/hair/pattern/mod.rs

```diff
@@ -1072,34 +1077,40 @@ pub fn compare_const_vals<'a, 'tcx>(
                 l.partial_cmp(&r)
             },
             ty::TyInt(_) => {
-                let a = interpret::sign_extend(tcx, a, ty).expect("layout error for TyInt");
-                let b = interpret::sign_extend(tcx, b, ty).expect("layout error for TyInt");
+                let a = interpret::sign_extend(tcx, a, ty.value).expect("layout error for TyInt");
+                let b = interpret::sign_extend(tcx, b, ty.value).expect("layout error for TyInt");
                 Some((a as i128).cmp(&(b as i128)))
             },
             _ => Some(a.cmp(&b)),
         }
     }
 
-    if let ty::TyRef(_, rty, _) = ty.sty {
+    if let ty::TyRef(_, rty, _) = ty.value.sty {
         if let ty::TyStr = rty.sty {
             match (a.to_byval_value(), b.to_byval_value()) {
                 (
-                    Some(Value::ByValPair(
-                        PrimVal::Ptr(ptr_a),
-                        PrimVal::Bytes(size_a))
-                    ),
-                    Some(Value::ByValPair(
-                        PrimVal::Ptr(ptr_b),
-                        PrimVal::Bytes(size_b))
-                    )
-                ) if size_a == size_b => {
-                    if ptr_a.offset == Size::from_bytes(0) && ptr_b.offset == Size::from_bytes(0) {
-                        let map = tcx.alloc_map.lock();
-                        let alloc_a = map.unwrap_memory(ptr_a.alloc_id);
-                        let alloc_b = map.unwrap_memory(ptr_b.alloc_id);
-                        if alloc_a.bytes.len() as u64 == size_a as u64 {
-                            return from_bool(alloc_a == alloc_b);
-                        }
+                    Some(Value::ScalarPair(
+                        Scalar::Ptr(ptr_a),
+                        Scalar::Bits {
+                            bits: size_a,
+                            defined: defined_a,
+                        },
+                    )),
+                    Some(Value::ScalarPair(
+                        Scalar::Ptr(ptr_b),
+                        Scalar::Bits {
+                            bits: size_b,
+                            defined: defined_b,
+                        },
+                    ))
+                ) if size_a == size_b && defined_a == defined_b &&
+                     ptr_a.offset.bytes() == 0 && ptr_b.offset.bytes() == 0 &&
+                     tcx.data_layout.pointer_size.bits() == defined_a.into() => {
```

Comment:
You're comparing `defined` again. At most these should `<=` relationships, not `==`.

Reply:
That is, each of `defined_a` and `defined_b` should be larger than `pointer_size.bits()` - why not use `to_bits` with `usize` instead?